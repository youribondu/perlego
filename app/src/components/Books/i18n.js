const translations = {
  fr: {
    endTitle: "Vous avez tous vus",
  },
  en: {
    endTitle: "You have seen it all",
  },
};

export default translations;
