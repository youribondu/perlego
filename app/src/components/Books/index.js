import React, { useEffect, useState, useContext } from "react";
import Book from "../Book";
import * as API from "../../services/api";
import { ApiRoutes } from "../../settings/routes";
import Skeleton from "react-loading-skeleton";
import InfiniteScroll from "react-infinite-scroll-component";
import "./index.scss";
import translations from "./i18n";
import LocaleContext from "../../services/context";
import { LOCALE_DEFAULT } from "../../settings/constants";

const Books = () => {
  const [books, setBooks] = useState([]);
  const [skip, setSkip] = useState(0);
  const [hasMore, setHasMore] = useState(false);

  const { locale } = useContext(LocaleContext) ?? LOCALE_DEFAULT;

  useEffect(() => {
    getBooks(skip);
  }, []);

  /**
   * Method to get books
   */
  const getBooks = (skip) => {
    // Call API to get books
    API.get(ApiRoutes.GET_BOOKS, { skip })
      .then(({ data }) => {
        setBooks([...books, ...data.results]);

        // Increment skip
        setSkip(skip + data.totalResults);

        // Define if there are more books to get
        setHasMore(data.hasMore);
      })
      .catch((error) => {
        // Init states
        setBooks([]);
        setSkip(0);
      });
  };

  const renderSkeletons = () => (
    <ul className="p-books">
      {[...Array(12)].map((value, index) => (
        <li key={index} className="p-books__book">
          <Skeleton height={180} />
          <Skeleton />
          <Skeleton />
        </li>
      ))}
    </ul>
  );

  return (
    <InfiniteScroll
      dataLength={skip}
      next={() => getBooks(skip)}
      hasMore={hasMore}
      loader={renderSkeletons()}
      endMessage={
        <p className="p-books__end">
          <b data-testid="books-end-message">{translations[locale].endTitle}</b>
        </p>
      }
    >
      <ul className="p-books">
        {books &&
          books.map((book, index) => (
            <li key={index} className="p-books__book">
              <Book {...book} />
            </li>
          ))}
      </ul>
    </InfiniteScroll>
  );
};

export default Books;
