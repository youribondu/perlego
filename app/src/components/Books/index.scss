@import "../../styles/index.scss";

.p-books {
  display: grid;
  gap: rem(15) calc(16.67% - 160px);
  width: 100%;
  grid-template-columns: repeat(6, minmax(0px, 1fr));

  &__book {
    @include card(true);
    margin: rem(10);
  }

  &__end {
    display: flex;
    justify-content: center;
    b {
      color: $color-primary;
    }
  }
}

@media only screen and (max-width: 600px) {
  .p-books {
    grid-template-columns: repeat(2, minmax(0px, 1fr));
  }
}
