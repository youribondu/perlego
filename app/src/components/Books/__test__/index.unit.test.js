import React from "react";
import ReactDom from "react-dom";
import Books from "../index";
import { cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { create } from "react-test-renderer";
import InfiniteScroll from "react-infinite-scroll-component";

const MockInfiniteScroll = () => {
  React.useEffect(() => {}, []);
  return (
    <div className="infinite-scroll-component__outerdiv" style={{}}>
      <div
        className="infinite-scroll-component "
        style={{
          WebkitOverflowScrolling: "touch",
          height: "auto",
          overflow: "auto",
        }}
      >
        <ul className="p-books" />
        <p className="p-books__end">
          <b data-testid="books-end-message">You have seen it all</b>
        </p>
      </div>
    </div>
  );
};

jest.mock("react-infinite-scroll-component", () => ({
  __esModule: true,
  namedExport: jest.fn(),
  default: jest.fn(),
}));

beforeAll(() => {
  InfiniteScroll.mockImplementation(MockInfiniteScroll);
});

describe("Books Component", () => {
  afterEach(cleanup);

  it("Renders without crashing", () => {
    const div = document.createElement("div");
    ReactDom.render(<Books />, div);
  });

  it("Matches snapshot empty", () => {
    const app = create(<Books />).toJSON();
    expect(app).toMatchSnapshot();
  });
});
