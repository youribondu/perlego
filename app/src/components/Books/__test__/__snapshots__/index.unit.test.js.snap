// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books Component Matches snapshot empty 1`] = `
<div
  className="infinite-scroll-component__outerdiv"
  style={Object {}}
>
  <div
    className="infinite-scroll-component "
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "height": "auto",
        "overflow": "auto",
      }
    }
  >
    <ul
      className="p-books"
    />
    <p
      className="p-books__end"
    >
      <b
        data-testid="books-end-message"
      >
        You have seen it all
      </b>
    </p>
  </div>
</div>
`;
